src/hotspot/share/runtime/javaThread.cpp

```diff
@@ -745,6 +745,7 @@ static void ensure_join(JavaThread* thread) {
   // Clear the native thread instance - this makes isAlive return false and allows the join()
   // to complete once we've done the notify_all below. Needs a release() to obey Java Memory Model
   // requirements.
+  guarantee(java_lang_Thread::thread(threadObj()) == thread, "must be alive");
```

Comment:
Please make this an assert rather than a guarantee. Only compiler threads can potentially violate this so we should not pay the price of the check for every thread .

Reply:
Ok.  It's not exactly an expensive test.